File: android/app/src/main/java/com/fghbuild/caic/voice/FunctionHandlers.kt
// Dispatches Gemini function calls to the caic API.
package com.fghbuild.caic.voice

import com.caic.sdk.v1.ApiClient
import com.caic.sdk.v1.CreateTaskReq
import com.caic.sdk.v1.EventKinds
import com.caic.sdk.v1.InputReq
import com.caic.sdk.v1.Prompt
import com.caic.sdk.v1.SyncReq
import com.caic.sdk.v1.Task
import com.fghbuild.caic.data.TaskRepository
import com.fghbuild.caic.data.TaskSSEEvent
import com.fghbuild.caic.util.formatCost
import com.fghbuild.caic.util.formatElapsed
import kotlinx.coroutines.flow.takeWhile
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonPrimitive

class FunctionHandlers(
    private val apiClient: ApiClient,
    private val taskRepository: TaskRepository,
    private val baseURL: String,
    private val taskNumberMap: TaskNumberMap,
    private val excludedTaskIds: () -> Set<String>,
) {

    suspend fun handle(name: String, args: JsonObject): JsonElement {
        return try {
            when (name) {
                "tasks_list" -> handleListTasks()
                "task_create" -> handleCreateTask(args)
                "task_get_detail" -> handleGetTaskDetail(args)
                "task_send_message" -> handleSendMessage(args)
                "task_answer_question" -> handleAnswerQuestion(args)
                "task_push_branch_to_remote" -> handleSyncTask(args)
                "task_terminate" -> handleTerminateTask(args)
                "get_usage" -> handleGetUsage()
                "list_repos" -> handleListRepos()
                "task_get_last_message_from_assistant" -> handleGetLastMessage(args)
                else -> errorResult("Unknown function: $name")
            }
        } catch (@Suppress("TooGenericExceptionCaught") e: Exception) {
            errorResult("Error: ${e.message}")
        }
    }

    private suspend fun handleListTasks(): JsonElement {
        val excluded = excludedTaskIds()
        val tasks = apiClient.listTasks().filter { it.id !in excluded }
        if (tasks.isEmpty()) return textResult("No tasks running.")
        val lines = tasks.joinToString("\n") { t ->
            val num = taskNumberMap.toNumber(t.id) ?: 0
            taskSummaryLine(num, t)
        }
        return textResult("## Tasks\n\n$lines")
    }

    private suspend fun handleCreateTask(args: JsonObject): JsonElement {
        val prompt = args.requireString("prompt")
        val repo = resolveRepo(args.requireString("repo"))
            ?: return errorResult("Unknown repo: ${args.requireString("repo")}")
        val model = args.optString("model")
        val harness = args.optString("harness") ?: "claude"
        val resp = apiClient.createTask(
            CreateTaskReq(
                initialPrompt = Prompt(text = prompt),
                repo = repo,
                model = model,
                harness = harness,
            )
        )
        // Refresh the map so the new task gets a number.
        val excluded = excludedTaskIds()
        val tasks = apiClient.listTasks().filter { it.id !in excluded }
        taskNumberMap.update(tasks)
        val num = taskNumberMap.toNumber(resp.id)
        return if (num != null) {
            textResult("Created task #$num: ${prompt.lines().first().take(SHORT_NAME_MAX)}")
        } else {
            textResult("Created task: ${prompt.lines().first().take(SHORT_NAME_MAX)}")
        }
    }

    private suspend fun handleGetTaskDetail(args: JsonObject): JsonElement {
        val taskId = resolveTaskNumber(args) ?: return errorResult("Unknown task number")
        val num = args.requireInt("task_number")
        val tasks = apiClient.listTasks()
        val t = tasks.find { it.id == taskId }
            ?: return errorResult("Task #$num not found")
        val shortName = t.initialPrompt.lines().firstOrNull()?.take(SHORT_NAME_MAX) ?: t.id
        val detail = buildString {
            appendLine("## Task #$num: $shortName")
            appendLine()
            append("**State:** ${t.state}  ")
            append("**Elapsed:** ${formatElapsed(t.duration)}  ")
            appendLine("**Cost:** ${formatCost(t.costUSD)}")
            when {
                t.state == "asking" -> appendLine("Waiting for user input before it can continue.")
                t.state == "terminated" && !t.result.isNullOrBlank() ->
                    appendLine("**Result:** ${t.result}")
                t.state == "failed" ->
                    appendLine("**Error:** ${t.error ?: "unknown"}")
            }
            t.diffStat?.takeIf { it.isNotEmpty() }?.let { diff ->
                append("**Changed:** ${diff.joinToString(", ") { it.path }}")
            }
        }.trim()
        return textResult(detail)
    }

    private suspend fun handleSendMessage(args: JsonObject): JsonElement {
        val taskId = resolveTaskNumber(args) ?: return errorResult("Unknown task number")
        val num = args.requireInt("task_number")
        val message = args.requireString("message")
        apiClient.sendInput(taskId, InputReq(prompt = Prompt(text = message)))
        return textResult("Sent message to task #$num.")
    }

    private suspend fun handleAnswerQuestion(args: JsonObject): JsonElement {
        val taskId = resolveTaskNumber(args) ?: return errorResult("Unknown task number")
        val num = args.requireInt("task_number")
        val answer = args.requireString("answer")
        apiClient.sendInput(taskId, InputReq(prompt = Prompt(text = answer)))
        return textResult("Answered task #$num.")
    }

    private suspend fun handleSyncTask(args: JsonObject): JsonElement {
        val taskId = resolveTaskNumber(args) ?: return errorResult("Unknown task number")
        val num = args.requireInt("task_number")
        val force = args["force"]?.jsonPrimitive?.booleanOrNull ?: false
        val target = args.optString("target").let { if (it == "main" || it == "master") "default" else it }
        val resp = apiClient.syncTask(taskId, SyncReq(force = force, target = target))
        val issues = resp.safetyIssues
        val verb = if (target == "default") "Pushed task #$num to main" else "Synced task #$num"
        return if (issues.isNullOrEmpty()) {
            textResult("$verb.")
        } else {
            val issueLines = issues.joinToString("\n") { "- **${it.kind}** ${it.file}: ${it.detail}" }
            textResult("$verb with safety issues:\n$issueLines")
        }
    }

    private suspend fun handleTerminateTask(args: JsonObject): JsonElement {
        val taskId = resolveTaskNumber(args) ?: return errorResult("Unknown task number")
        val num = args.requireInt("task_number")
        apiClient.terminateTask(taskId)
        return textResult("Terminated task #$num.")
    }

    private suspend fun handleGetUsage(): JsonElement {
        val usage = apiClient.getUsage()
        fun pct(v: Double) = "${v.toInt()}%"
        val summary = buildString {
            appendLine("5-hour window: ${pct(usage.fiveHour.utilization)} used, resets ${usage.fiveHour.resetsAt}")
            append("7-day window: ${pct(usage.sevenDay.utilization)} used, resets ${usage.sevenDay.resetsAt}")
            if (usage.extraUsage.isEnabled) {
                appendLine()
                val usedDollars = usage.extraUsage.usedCredits / 100
                val limitDollars = usage.extraUsage.monthlyLimit / 100
                append(
                    "Extra usage: \$${usedDollars.toInt()} of " +
                        "\$${limitDollars.toInt()} monthly limit used",
                )
            }
        }
        return textResult(summary)
    }

    private suspend fun handleGetLastMessage(args: JsonObject): JsonElement {
        val taskId = resolveTaskNumber(args) ?: return errorResult("Unknown task number")
        val num = args.requireInt("task_number")

        val events = mutableListOf<com.caic.sdk.v1.ClaudeEventMessage>()
        taskRepository.taskRawEventsWithReady(baseURL, taskId)
            .takeWhile { it !is TaskSSEEvent.Ready }
            .collect { event ->
                if (event is TaskSSEEvent.Event) events.add(event.msg)
            }

        val message = events.lastOrNull { it.kind == EventKinds.Result }?.result?.result?.let { r ->
            "Task #$num result: $r"
        } ?: events.lastOrNull { it.kind == EventKinds.Ask }?.ask?.questions?.firstOrNull()?.let { q ->
            val opts = q.options.joinToString(", ") { it.label }
            "Task #$num is asking: ${q.question} Options: $opts"
        } ?: events.lastOrNull { it.kind == EventKinds.Text }?.text?.text?.let { t ->
            "Last message from task #$num: $t"
        } ?: "No messages from task #$num yet."
        return textResult(message)
    }

    private suspend fun handleListRepos(): JsonElement {
        val repos = apiClient.listRepos()
        if (repos.isEmpty()) return textResult("No repositories available.")
        val lines = repos.joinToString("\n") { r ->
            "- **${r.path}** (base: ${r.baseBranch})"
        }
        return textResult("## Repositories\n\n$lines")
    }

    /** Resolve a repo name to its canonical path using case-insensitive matching. */
    private suspend fun resolveRepo(name: String): String? {
        val repos = apiClient.listRepos()
        return repos.find { it.path.equals(name, ignoreCase = true) }?.path
    }

    /** Resolve task_number from args to a real task ID via the map. */
    private fun resolveTaskNumber(args: JsonObject): String? {
        val num = args.requireInt("task_number")
        return taskNumberMap.toId(num)
    }
}

private fun taskSummaryLine(num: Int, t: Task): String {
    val name = t.initialPrompt.lines().firstOrNull()?.take(SHORT_NAME_MAX) ?: t.id
    val base = "$num. **$name** — ${t.state}, ${formatElapsed(t.duration)}, " +
        "${formatCost(t.costUSD)}, ${t.harness}"
    return when {
        t.state == "asking" -> "$base — NEEDS INPUT"
        t.state == "terminated" && !t.result.isNullOrBlank() ->
            "$base — Result: ${t.result!!.take(RESULT_SNIPPET_MAX)}"
        t.state == "failed" -> "$base — Error: ${t.error ?: "unknown"}"
        else -> base
    }
}

private const val SHORT_NAME_MAX = 40
private const val RESULT_SNIPPET_MAX = 120

private fun JsonObject.requireString(key: String): String =
    this[key]?.jsonPrimitive?.content
        ?: throw IllegalArgumentException("Missing required parameter: $key")

private fun JsonObject.requireInt(key: String): Int =
    this[key]?.jsonPrimitive?.intOrNull
        ?: throw IllegalArgumentException("Missing required integer parameter: $key")

private fun JsonObject.optString(key: String): String? =
    this[key]?.jsonPrimitive?.content

private fun textResult(message: String): JsonElement =
    JsonObject(mapOf("result" to JsonPrimitive(message)))

private fun errorResult(message: String): JsonElement =
    JsonObject(mapOf("error" to JsonPrimitive(message)))

File: frontend/src/FunctionHandlers.ts
// Gemini function call dispatch for voice mode,
// parallel to android/voice/FunctionHandlers.kt.
import {
  listTasks,
  createTask,
  sendInput,
  syncTask,
  terminateTask,
  getUsage,
  cloneRepo,
  taskEvents,
  webFetch,
} from "./api";
import type { Task, EventMessage } from "@sdk/types.gen";
import { formatCost, formatElapsed } from "./formatting";
import type { TaskNumberMap } from "./TaskNumberMap";

type FunctionArgs = Record<string, unknown>;

function requireString(args: FunctionArgs, key: string): string {
  const v = args[key];
  if (typeof v !== "string") throw new Error(`Missing required parameter: ${key}`);
  return v;
}

function requireInt(args: FunctionArgs, key: string): number {
  const v = args[key];
  if (typeof v !== "number") throw new Error(`Missing required integer: ${key}`);
  return Math.floor(v);
}

function optString(args: FunctionArgs, key: string): string | undefined {
  const v = args[key];
  return typeof v === "string" ? v : undefined;
}

function optBool(args: FunctionArgs, key: string): boolean {
  return args[key] === true;
}

function textResult(message: string): Record<string, string> {
  return { result: message };
}

function errorResult(message: string): Record<string, string> {
  return { error: message };
}

const RESULT_SNIPPET_MAX = 120;

function diffStatSummary(t: Task): string {
  const ds = t.diffStat;
  if (!ds?.length) return "";
  let added = 0;
  let deleted = 0;
  for (const f of ds) {
    added += f.added;
    deleted += f.deleted;
  }
  return `, +${added} -${deleted} in ${ds.length} ${ds.length === 1 ? "file" : "files"}`;
}

function taskSummaryLine(num: number, t: Task): string {
  const name = t.title || t.id;
  const base = `${num}. **${name}** — ${t.state}, ${formatElapsed(t.duration * 1000)}, ${formatCost(t.costUSD)}, ${t.harness}${diffStatSummary(t)}`;
  if (t.state === "terminated" && t.result) return `${base} — ${t.result.slice(0, RESULT_SNIPPET_MAX)}`;
  if (t.state === "failed" && t.error) return `${base} — ${t.error}`;
  return base;
}

export class FunctionHandlers {
  constructor(
    private readonly taskNumberMap: TaskNumberMap,
    private readonly excludedTaskIds: () => Set<string>,
    private readonly defaultHarness = "",
    private readonly defaultModel = "",
  ) {}

  async handle(name: string, args: FunctionArgs): Promise<Record<string, unknown>> {
    try {
      switch (name) {
        case "tasks_list":
          return await this.handleListTasks();
        case "task_create":
          return await this.handleCreateTask(args);
        case "task_get_detail":
          return await this.handleGetTaskDetail(args);
        case "task_send_message":
          return await this.handleSendMessage(args);
        case "task_answer_question":
          return await this.handleAnswerQuestion(args);
        case "task_push_branch_to_remote":
          return await this.handleSyncTask(args);
        case "task_terminate":
          return await this.handleTerminateTask(args);
        case "get_usage":
          return await this.handleGetUsage();
        case "clone_repo":
          return await this.handleCloneRepo(args);
        case "task_get_last_message_from_assistant":
          return await this.handleGetLastMessage(args);
        case "web_search":
          return await this.handleWebSearch(args);
        case "web_fetch":
          return await this.handleWebFetch(args);
        default:
          return errorResult(`Unknown function: ${name}`);
      }
    } catch (e: unknown) {
      return errorResult(e instanceof Error ? e.message : "Unknown error");
    }
  }

  private async handleListTasks(): Promise<Record<string, unknown>> {
    const excluded = this.excludedTaskIds();
    const tasks = (await listTasks()).filter((t) => !excluded.has(t.id));
    if (tasks.length === 0) return textResult("No tasks running.");
    const lines = tasks.map((t) => taskSummaryLine(this.taskNumberMap.toNumber(t.id) ?? 0, t));
    return textResult(`## Tasks\n\n${lines.join("\n")}`);
  }

  private async handleCreateTask(args: FunctionArgs): Promise<Record<string, unknown>> {
    const prompt = requireString(args, "prompt");
    const repo = requireString(args, "repo");
    const model = optString(args, "model") ?? (this.defaultModel || undefined);
    const harness = optString(args, "harness") ?? this.defaultHarness;
    const resp = await createTask({
      initialPrompt: { text: prompt },
      repo,
      harness,
      ...(model ? { model } : {}),
    });
    const excluded = this.excludedTaskIds();
    const tasks = (await listTasks()).filter((t) => !excluded.has(t.id));
    this.taskNumberMap.update(tasks);
    const num = this.taskNumberMap.toNumber(resp.id);
    const title = tasks.find((t) => t.id === resp.id)?.title || resp.id;
    return textResult(num !== undefined ? `Created task #${num}: ${title}` : `Created task: ${title}`);
  }

  private async handleGetTaskDetail(args: FunctionArgs): Promise<Record<string, unknown>> {
    const num = requireInt(args, "task_number");
    const taskId = this.taskNumberMap.toId(num);
    if (!taskId) return errorResult("Unknown task number");
    const tasks = await listTasks();
    const t = tasks.find((x) => x.id === taskId);
    if (!t) return errorResult(`Task #${num} not found`);
    const shortName = t.title || t.id;
    const lines = [
      `## Task #${num}: ${shortName}`,
      "",
      `State: ${t.state}  Elapsed: ${formatElapsed(t.duration * 1000)}  Cost: ${formatCost(t.costUSD)}`,
    ];
    if (t.state === "terminated" && t.result) lines.push(`**Result:** ${t.result}`);
    if (t.state === "failed" && t.error) lines.push(`**Error:** ${t.error}`);
    if (t.diffStat?.length) lines.push(`**Changed:** ${t.diffStat.map((d) => d.path).join(", ")}`);
    return textResult(lines.join("\n").trim());
  }

  private async handleSendMessage(args: FunctionArgs): Promise<Record<string, unknown>> {
    const num = requireInt(args, "task_number");
    const taskId = this.taskNumberMap.toId(num);
    if (!taskId) return errorResult("Unknown task number");
    await sendInput(taskId, { prompt: { text: requireString(args, "message") } });
    return textResult(`Sent message to task #${num}.`);
  }

  private async handleAnswerQuestion(args: FunctionArgs): Promise<Record<string, unknown>> {
    const num = requireInt(args, "task_number");
    const taskId = this.taskNumberMap.toId(num);
    if (!taskId) return errorResult("Unknown task number");
    await sendInput(taskId, { prompt: { text: requireString(args, "answer") } });
    return textResult(`Answered task #${num}.`);
  }

  private async handleSyncTask(args: FunctionArgs): Promise<Record<string, unknown>> {
    const num = requireInt(args, "task_number");
    const taskId = this.taskNumberMap.toId(num);
    if (!taskId) return errorResult("Unknown task number");
    const force = optBool(args, "force");
    const targetRaw = optString(args, "target");
    const target = targetRaw === "main" || targetRaw === "master" ? "default" : targetRaw;
    const resp = await syncTask(taskId, {
      ...(force ? { force } : {}),
      ...(target ? { target } : {}),
    });
    const verb = target === "default" ? `Pushed task #${num} to main` : `Synced task #${num}`;
    if (!resp.safetyIssues?.length) return textResult(`${verb}.`);
    const issueLines = resp.safetyIssues
      .map((i) => `- **${i.kind}** ${i.file}: ${i.detail}`)
      .join("\n");
    return textResult(`${verb} with safety issues:\n${issueLines}`);
  }

  private async handleTerminateTask(args: FunctionArgs): Promise<Record<string, unknown>> {
    const num = requireInt(args, "task_number");
    const taskId = this.taskNumberMap.toId(num);
    if (!taskId) return errorResult("Unknown task number");
    await terminateTask(taskId);
    return textResult(`Terminated task #${num}.`);
  }

  private async handleGetUsage(): Promise<Record<string, unknown>> {
    const usage = await getUsage();
    const pct = (v: number) => `${Math.floor(v)}%`;
    const lines = [
      `5-hour window: ${pct(usage.fiveHour.utilization)} used, resets ${usage.fiveHour.resetsAt}`,
      `7-day window: ${pct(usage.sevenDay.utilization)} used, resets ${usage.sevenDay.resetsAt}`,
    ];
    if (usage.extraUsage.isEnabled) {
      const usedDollars = Math.floor(usage.extraUsage.usedCredits / 100);
      const limitDollars = Math.floor(usage.extraUsage.monthlyLimit / 100);
      lines.push(`Extra usage: $${usedDollars} of $${limitDollars} monthly limit used`);
    }
    return textResult(lines.join("\n"));
  }

  private async handleCloneRepo(args: FunctionArgs): Promise<Record<string, unknown>> {
    const url = requireString(args, "url");
    const path = optString(args, "path");
    const repo = await cloneRepo({ url, ...(path ? { path } : {}) });
    return textResult(`Cloned **${repo.path}** (base: ${repo.baseBranch}).`);
  }

  private async handleWebSearch(args: FunctionArgs): Promise<Record<string, unknown>> {
    const query = requireString(args, "query");
    const url = `https://html.duckduckgo.com/html/?q=${encodeURIComponent(query)}`;
    const resp = await webFetch({ url });
    return { title: resp.title, content: resp.content };
  }

  private async handleWebFetch(args: FunctionArgs): Promise<Record<string, unknown>> {
    const url = requireString(args, "url");
    const resp = await webFetch({ url });
    return { title: resp.title, content: resp.content };
  }

  private handleGetLastMessage(args: FunctionArgs): Promise<Record<string, unknown>> {
    const num = requireInt(args, "task_number");
    const taskId = this.taskNumberMap.toId(num);
    if (!taskId) return Promise.resolve(errorResult("Unknown task number"));

    return new Promise((resolve) => {
      const collected: EventMessage[] = [];
      let settled = false;
      let historyTimer: ReturnType<typeof setTimeout> | null = null;

      const settle = () => {
        if (settled) return;
        settled = true;
        if (historyTimer !== null) clearTimeout(historyTimer);
        es.close();
        const lastResult = [...collected].reverse().find((e) => e.kind === "result");
        const lastAsk = [...collected].reverse().find((e) => e.kind === "ask");
        const lastText = [...collected].reverse().find((e) => e.kind === "text");
        if (lastResult?.result?.result) {
          resolve(textResult(`Task #${num} result: ${lastResult.result.result}`));
        } else if (lastAsk?.ask?.questions?.[0]) {
          const q = lastAsk.ask.questions[0];
          const opts = q.options.map((o) => o.label).join(", ");
          resolve(
            textResult(
              `Task #${num} is asking: ${q.question}${opts ? ` Options: ${opts}` : ""}`,
            ),
          );
        } else if (lastText?.text?.text) {
          resolve(textResult(`Last message from task #${num}: ${lastText.text.text}`));
        } else {
          resolve(textResult(`No messages from task #${num} yet.`));
        }
      };

      const es = taskEvents(taskId, (event) => {
        collected.push(event);
        if (event.kind === "init") {
          // Wait 1s after init to let the replay stream in.
          if (historyTimer !== null) clearTimeout(historyTimer);
          historyTimer = setTimeout(settle, 1000);
        }
      });

      // Fallback: settle after 5s regardless.
      setTimeout(settle, 5000);
    });
  }
}
